pubsub.go

```diff
@@ -163,6 +185,101 @@ func (p *PubsubValueStore) GetValue(ctx context.Context, key string, opts ...rop
 	return p.getLocal(key)
 }
 
+func (wg *watchGroup) run(p *PubsubValueStore, key string) {
+	for {
+		select {
+		case w := <-wg.register:
+			wg.listeners[w.out] = w.ctx
+
+			//ack with dummy message
+			w.out <- []byte{}
+		case w := <-wg.unregister:
+			delete(wg.listeners, w[0])
+
+			w[1] <- []byte{}
+		case <-wg.closing:
+			return
+		}
+	}
+}
+
+func (p *PubsubValueStore) SearchValue(ctx context.Context, key string, opts ...ropts.Option) (<-chan []byte, error) {
+	if err := p.Subscribe(key); err != nil {
+		return nil, err
+	}
+
+	out := make(chan []byte, 1)
+	lv, err := p.getLocal(key)
+	if err == nil {
+		out <- lv
+	}
+
+	p.watchLk.Lock()
+	wg, ok := p.watching[key]
+	if !ok {
+		wg = &watchGroup{
+			register: make(chan watcher),
+			unregister: make(chan []chan<- []byte),
+
+			closing:   make(chan struct{}),
+			listeners: map[chan<- []byte]context.Context{},
+		}
+		p.watching[key] = wg
+		go wg.run(p, key)
+	}
+
+	proxy := make(chan []byte, 1)
+
+	ctx, cancel := context.WithCancel(ctx)
+	wg.register <- watcher{ctx: ctx, out: proxy}
+
+	// we wait for ack from watchgroup manager, essentially transferring our lock
```

Comment:
We *really* shouldn't do this. Implicitly transferring a lock is guaranteed to trip someone up down the road.

At this point, we effectively have one big lock *anyways* (as far as I can tell), we might as well just use it and drop the watch manager.